### DISCUSSION TITLE
Some 1-x practicals appear not to be interactive

### DISCUSSION THREAD

**Discussion by richardreeve** [IBAHCM/RPiR#122] [Archive] [answered]
Some people are finding practicals, especially 1-2, appear as a series of non-interactive R chunks with no ability to edit them. Suggestions for fixing this?

**Comment by soniamitchell** [comment]
Yes, Practical 1-2 onwards will look like this if they used `install_github("IBAHCM/RPiR", dependencies = FALSE)`. The problem is, they don't have their dependencies installed...

**Reply by richardreeve** [reply]
That would make sense! I've made a potential series of solutions below - see whether you think I've missed some.

**Comment by richardreeve** [comment, marked as answer]
This seems to be caused by installation errors of some kind. There are a lot of possible solutions that get increasingly extreme:

1. If you actually get an error, then see [here](https://ibahcm.github.io/RPiR/articles/pages/install_RPiR.html#potential-problems) for suggestions 
2. If not, try running `RPiR:::clean_cached_exercises()` and then restart R and try again
3. If that doesn't work, try restarting R, reinstalling the `rmarkdown` package (with `install.packages("rmarkdown")`), then try reloading the `RPiR` library and running the practical again
4. Try restarting R, reinstalling the `RPiR` package (with `library(devtools)`, `install_github("IBAHCM/RPiR", force=TRUE)`), then try reloading the `RPiR` library and running the practical again
5. Try restarting R, uninstalling `RPiR` (with `remove.packages("RPiR")`) and then 4 (above) again

If at any point you are installing RPiR and you get any errors, follow the instructions [here](https://ibahcm.github.io/RPiR/articles/pages/install_RPiR.html). If you get repeated errors that you can't remove or reinstall a package, try closing all of your RStudio instances and restart what you were doing.

**Reply by richardreeve** [reply]
Incidentally, 3. is especially likely to work if as well as not being able to run a practical, you also can't run a lecture properly (e.g. using `run_lecture("RPiR1-intro")`).

**Reply by soniamitchell** [reply]
I would emphasise point 1 above.. particularly [Try removing ‘C:.../00LOCK’](https://ibahcm.github.io/RPiR/articles/pages/install_RPiR.html#try-removing-c---00lock). If there is no 00LOCK file at the location, please take a screenshot. I'm hesitant to extend the troubleshooting instructions without pictures, since it involves deleting directories.

**Reply by richardreeve** [reply]
I think that's fine - people should contact us at one of the classes or through this discussion forum if these don't fix things, and we can add extra solutions here.

**Reply by soniamitchell** [reply]
copying this across from the other thread for completeness

> Just to emphasise, please don't use install_github("IBAHCM/RPiR", dependencies = FALSE). If you've done this, you might not have the correct dependencies installed and Practical 1-2 will appear broken. To fix this, use install_github("IBAHCM/RPiR", force = TRUE) and select 1 to install all dependencies.

**Comment by soniamitchell** [comment]
I've found an issue on the {learnr} package [repo](https://github.com/rstudio/learnr/issues/344). They seem to think it might be a pandoc issue. If I run `rmarkdown::pandoc_version()` I get 2.3.1, what do you get @shayavh? You might need to install the newest version of [pandoc](https://pandoc.org/installing.html)..

**Reply by shayavh** [reply]
I get 2.11.2

**Reply by soniamitchell** [reply]
I guess it's not that then! Though others seem to be having this problem... so hopefully there'll be an answer soon https://github.com/rstudio/learnr/issues/419

**Reply by soniamitchell** [reply]
@shayavh if you get some free time... could you try, in Rstudio, View > Panes > Pane Layout... then in the [Files, Plots, Packages, Help] box, scroll down and check Tutorial. Then try launching tutorials from there... (after a clean install and restart)

If that doesn't work then `find.package("RPiR")` will find where RPiR is installed. Then if you find that location on your computer, then go to tutorials and the appropriate practical, and delete the *.html file, then run RPiR::run_practical(), that might avoid the need to reinstall... not sure which you think is easier...

I also seem to have woken up the [thread](https://github.com/rstudio/learnr/issues/419).. not sure if we want to add our experiences over there @richardreeve? If we can provide a print out of @shayavh 's package versions and system environment they might be able to reproduce the error? maybe?

**Reply by shayavh** [reply]
@soniamitchell it now gives me errors when trying to install RPiR again, but they are different:

ERROR: failed to lock directory ‘/Library/Frameworks/R.framework/Versions/4.0/Resources/library’ for modifying
Try removing ‘/Library/Frameworks/R.framework/Versions/4.0/Resources/library/00LOCK-RPiR’

Error: Failed to install 'RPiR' from GitHub:
  (converted from warning) installation of package ‘/var/folders/m0/9m8ntcfd7_1c972zp1480j1h0000gn/T//RtmpORADiA/file6781622080/RPiR_0.41.3.tar.gz’ had non-zero exit status

**Reply by richardreeve** [reply]
Ah, well, that's fine (for some definition of fine, sorry!) - it's another error that R generates detailed in the link from 1, above, and we know how to fix it. Basically, this is a known, old, problem that will crop up from time to time and you just need to follow the instructions and find */Library/Frameworks/R.framework/Versions/4.0/Resources/library/00LOCK-RPiR* on your computer and delete it.

**Reply by richardreeve** [reply]
@soniamitchell:

> If that doesn't work then find.package("RPiR") will find where RPiR is installed. Then if you find that location on your computer, then go to tutorials and the appropriate practical, and delete the *.html file, then run RPiR::run_practical(), that might avoid the need to reinstall... not sure which you think is easier...

All we need to do here is to write a function that deletes all of the html files in the tutorial folders. That should be relatively simple once we work out where they are stored... if that works we're golden. It explains why reinstalling `RPiR` helps too, because it wipes those files. It doesn't explain why reinstalling `rmarkdown` often helps, but maybe some html files get cached there too sometimes. Anyway, this is really good news. I'll have a look at it tomorrow, but it should be a pretty easy fix... basically *if you get this problem, just run the `clean_tutorial_caches()` function, and everything should be fine again!* Anyway, I'll have a look at this tomorrow. Too late today!

**Reply by soniamitchell** [reply]
Good idea! Though supposedly, using the Tutorial pane in RStudio avoids the problem altogether.. which is even better! (until the underlying bug is found in learnr / rmarkdown / pandoc)

**Reply by soniamitchell** [reply]
Hi @shayavh, did you manage to delete the 00LOCK-RPiR file, `install_github()`, and get the Tutorial pane up?

**Reply by shayavh** [reply]
Hi @soniamitchell, no I don't think the directory that it states exists on my computer. And the whole RPiR package has disappeared from my packages as well, so I can't find it anywhere.

**Reply by soniamitchell** [reply]
Hey @shayavh, you have a mac don't you? You need to navigate all the way back to the base of your file system, in my case MacBook Pro. If you open any folder (see my example below), right click on the title of the window (in my example, a little blue folder next to "library"), and a menu should appear (see the first picture below). Go all the way down to the bottom and start navigating from there... Alternatively, the name of your laptop might appear in the navigational pane on the left of any folder you open (not sure what this window is called). If OOLOCK doesn't exist, just delete the RPiR directory. Then run `install_github()`.. and then try to find the Tutorial Pane.

<img width="176" alt="Screenshot 2021-01-23 at 11 31 48" src="https://user-images.githubusercontent.com/8394297/105577073-3791ff00-5d6f-11eb-8d79-563f2a34bd05.png">
<img width="882" alt="Screenshot 2021-01-23 at 11 33 09" src="https://user-images.githubusercontent.com/8394297/105577075-3a8cef80-5d6f-11eb-82e4-420948417fb4.png">

**Reply by shayavh** [reply]
ah thanks @soniamitchell, I deleted it and am now able to install. I also checked the tutorials in panes. How do I open it with tutorials?

**Reply by soniamitchell** [reply]
What does your panel look like @shayavh ? Mine is already populated... maybe try `library(RPiR)`? 
<img width="1461" alt="Screenshot 2021-01-23 at 14 11 14" src="https://user-images.githubusercontent.com/8394297/105580555-f3115e00-5d84-11eb-887d-327b0fa7dc44.png">

**Reply by shayavh** [reply]
Ah yes I see it now @soniamitchell . It still only brings up one practical at a time though. So like 0-1 will have the boxes, but 1-0 won't. But it is fine though, I can just install R-markdown for that and we don't need it anymore from next week. Thanks again!

**Reply by soniamitchell** [reply]
Ahh.. that's annoying. It's a shame I can't play with your laptop to see what's going on.. but you're right, you don't need it from next week... the practicals are now online https://ibahcm.github.io/RPiR/ 🚀🎉 @shayavh

**Reply by richardreeve** [reply]
@shayavh If you update the RPiR package again, there is a now function that you can call using:

```R
RPiR:::clean_cached_exercises()
```

Note the three colons. Hopefully it will wipe the corrupted files. Could you try to run a couple of exercises until you get this problem again, and then restart R and try again (without doing any reinstallation) and see if it helps?

**Reply by shayavh** [reply]
@richardreeve Restarting R, then library(RPiR), then RPiR:::clean_cached_exercises() before every practical works. But I can't run the practicals one after another without doing this in between.
